### DISCUSSION TITLE
Help configuring SMTP

### DISCUSSION THREAD

**Discussion by kennylajara** [dani-garcia/vaultwarden#1711] [Q&A] [answered]
I can't make the SMTP works. This is my config:

```
Enabled:  true/checked
Host:   smtp.gmail.com
Enable Secure SMTP:   true/checked
Force TLS:   false/unchecked
Port:   465
From Address:   myaccount@gmail.com
From Name:   Notifications
Username:   myaccount@gmail.com
Password:   ••••••••••
SMTP Auth mechanism:   [empty]
SMTP connection timeout:   15
Server name sent during HELO:   [empty]
Enable SMTP debugging:   false
Accept Invalid Certs:   false
Accept Invalid Hostnames:   false
Test SMTP:   myaccount@gmail.com
```

Also, I activated the GMail option for less secure apps at gmail: https://myaccount.google.com/lesssecureapps

**Error Messsage**

> Error sending SMTP test email
> SmtpError

**Server Logs**

> 2021-05-18 14:05:08.949][request][INFO] POST /admin/test/smtp/
> [2021-05-18 14:05:18.998][error][ERROR] SmtpError.
> [CAUSE] lettre::transport::smtp::Error {
>     kind: Response,
>     source: "incomplete response",
> }
> [2021-05-18 14:05:18.998][response][INFO] POST /admin/test/smtp (test_smtp) => 400 Bad Request

### Your environment (Generated via diagnostics page)
* Vaultwarden version: v1.21.0
* Web-vault version: v2.19.0d
* Running within Docker: true
* Uses a reverse proxy: true
* IP Header check: true (X-Real-IP)
* Internet access: true
* Internet access via a proxy: false
* DNS Check: true
* Time Check: true
* Domain Configuration Check: true
* HTTPS Check: true
* Database type: SQLite
* Database version: 3.33.0
* Clients used: 
* Reverse proxy and version: 
* Other relevant information: 

### Config (Generated via diagnostics page)
```json
{
  "_duo_akey": null,
  "_enable_duo": false,
  "_enable_email_2fa": false,
  "_enable_smtp": true,
  "_enable_yubico": false,
  "_ip_header_enabled": true,
  "admin_token": "***",
  "allowed_iframe_ancestors": "",
  "attachments_folder": "data/attachments",
  "authenticator_disable_time_drift": false,
  "data_folder": "data",
  "database_max_conns": 10,
  "database_url": "****/**.*******",
  "db_connection_retries": 15,
  "disable_2fa_remember": false,
  "disable_admin_token": false,
  "disable_icon_download": false,
  "domain": "*****://****.******.***",
  "domain_origin": "*****://****.******.***",
  "domain_path": "",
  "domain_set": true,
  "duo_host": null,
  "duo_ikey": null,
  "duo_skey": null,
  "email_attempts_limit": 3,
  "email_expiration_time": 600,
  "email_token_size": 6,
  "enable_db_wal": true,
  "extended_logging": true,
  "helo_name": null,
  "hibp_api_key": null,
  "icon_blacklist_non_global_ips": true,
  "icon_blacklist_regex": null,
  "icon_cache_folder": "data/icon_cache",
  "icon_cache_negttl": 259200,
  "icon_cache_ttl": 2592000,
  "icon_download_timeout": 10,
  "invitation_org_name": "Vaultwarden",
  "invitations_allowed": true,
  "ip_header": "X-Real-IP",
  "job_poll_interval_ms": 30000,
  "log_file": null,
  "log_level": "Info",
  "log_timestamp_format": "%Y-%m-%d %H:%M:%S.%3f",
  "org_attachment_limit": null,
  "org_creation_users": "",
  "password_iterations": 100003,
  "reload_templates": false,
  "require_device_email": false,
  "rsa_key_filename": "data/rsa_key",
  "send_purge_schedule": "0 5 * * * *",
  "sends_folder": "data/sends",
  "show_password_hint": false,
  "signups_allowed": true,
  "signups_domains_whitelist": "******.***",
  "signups_verify": false,
  "signups_verify_resend_limit": 6,
  "signups_verify_resend_time": 3600,
  "smtp_accept_invalid_certs": false,
  "smtp_accept_invalid_hostnames": false,
  "smtp_auth_mechanism": null,
  "smtp_debug": false,
  "smtp_explicit_tls": false,
  "smtp_from": "******.***********@*****.***",
  "smtp_from_name": "Vaultwarden Notifications",
  "smtp_host": "****.*****.***",
  "smtp_password": "***",
  "smtp_port": 465,
  "smtp_ssl": true,
  "smtp_timeout": 15,
  "smtp_username": "******.***********@*****.***",
  "templates_folder": "data/templates",
  "trash_auto_delete_days": 15,
  "trash_purge_schedule": "0 5 0 * * *",
  "use_syslog": false,
  "user_attachment_limit": null,
  "web_vault_enabled": true,
  "web_vault_folder": "web-vault/",
  "websocket_address": "0.0.0.0",
  "websocket_enabled": false,
  "websocket_port": 3012,
  "yubico_client_id": null,
  "yubico_secret_key": null,
  "yubico_server": null
}
```

**Comment by kennylajara** [comment]
I also tried with Sengrid and other mailserver and I am getting the same error for the 3 providers.

**Comment by kennylajara** [comment]
Now I tried to use Sendgrid and a I am getting the same error message (`SmtpError`) but with different logs. This is the log now:

> [2021-05-19 02:10:28.710][request][INFO] POST /admin/test/smtp/
> [2021-05-19 02:10:38.918][error][ERROR] SmtpError.
> [CAUSE] lettre::transport::smtp::Error {
>     kind: Network,
>     source: Os {
>         code: 11,
>         kind: WouldBlock,
>         message: "Resource temporarily unavailable",
>     },
> }
> [2021-05-19 02:10:38.918][response][INFO] POST /admin/test/smtp (test_smtp) => 400 Bad Request

**Comment by BlackDex** [comment, marked as answer]
The first error looks like you got a connection but it got cut or something. The last one looks like it isn't able to connect at all.

Try to enable `SMTP_DEBUG=true` to get some more information. In the logs

**Reply by kennylajara** [reply]
Thank you for the tip.

I didn't get more details with Sengrid by adding the `SMTP_DEBUG=true` option but I got some information with GMail:

> Host: smtp.gmail.com
> Enable Secure SMTP: True
> Force TLS: False
> Port: 587
> From Address: myaccount@gmail.com
> From Name: My Name
> Username: myaccount@gmail.com
> Password: SecretPassword123 # Not my real password
> SMTP Auth mechanism: `Empty`
> SMTP connection timeout: 15
> Server name sent during HELO: `Empty`

**Error Message:**

```
[2021-05-19 14:10:08.890][request][INFO] POST /admin/test/smtp/
[2021-05-19 14:10:09.319][vaultwarden::mail][ERROR] SMTP 5xx error: lettre::transport::smtp::Error { kind: Permanent(Code { severity: PermanentNegativeCompletion, category: Unspecified3, detail: Five }) }
[2021-05-19 14:10:09.319][response][INFO] POST /admin/test/smtp (test_smtp) => 400 Bad Request
```

I am not sure if I have to report this as a bug.

**Reply by BlackDex** [reply]
You should get more information then only that when having SMTP_DEBUG=true.
Are you sure you have restarted the container after setting that config, or pressed save after changing the settings in the admin interface?

**Reply by kennylajara** [reply]
I didn't know I had to restart the container after applying the changes (maybe it can be useful to inform it when you clic save).

Thanks to that information I was able to solve the issue. Thanks.

**Reply by BlackDex** [reply]
I have to make that a non editable config option.
Thx for letting us known.

**Reply by BlackDex** [reply]
And great to see it solved your issue

**Comment by iamLazyCode** [comment]
can anyone tell me where will i find these config to edit please , cant find anywhere

**Reply by BlackDex** [reply]
Did you checked the wiki?
https://github.com/dani-garcia/vaultwarden/wiki/Configuration-overview

**Reply by iamLazyCode** [reply]
thankyou sir i just found this and did the setup

https://github.com/dani-garcia/vaultwarden/wiki/SMTP-Configuration
